talktorials/11_online_apis/11c_quality_assessment.ipynb

```diff
@@ -0,0 +1,332 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_viewer(protein, ligands, affinities):\n",
+    "    \"\"\"\n",
+    "    Create a nglview widget with the protein and all the ligands labeled by affinities\n",
+    "    \"\"\"\n",
+    "    v = nv.show_file(protein)\n",
+    "    label_kwargs = dict(labelType=\"text\", sele=\"@0\", showBackground=True, backgroundColor=\"black\")\n",
+    "    for ligand, affinity in zip(ligands, affinities):\n",
+    "        c = v.add_component(ligand)\n",
+    "        c.add_label(labelText=[str(affinity)], **label_kwargs)\n",
+    "    return v"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this cell below we will build the actual GUI!\n",
+    "\n",
+    "It will be composed of two widgets arranged horizontally using the `AppLayout` layour.\n",
+    "\n",
+    "- The selector (`ipywidgets.Select`)\n",
+    "- The NGL viewer\n",
+    "\n",
+    "When the user clicks on a new entry in the selector, `_on_selection_change` will be called, which will:\n",
+    "\n",
+    "1. Check if the new value is any different from the previous one. If that's the case, then:\n",
+    "2. Hide all ligands (simpler way to hide the previous one; no need to check individually)\n",
+    "3. Show the new one and center the camera on it with a cool 500ms animation\n",
+    "4. Execute some JavaScript on the canvas to update the list of sidechains within 5A of the new pose center of mass."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 107,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# JavaScript code needed to update residues around the ligand\n",
+    "# because this part is not exposed in the Python widget\n",
+    "# Based on: http://nglviewer.org/ngl/api/manual/snippets.html\n",
+    "_RESIDUES_AROUND = \"\"\"\n",
+    "var protein = this.stage.compList[0];\n",
+    "var ligand_center = this.stage.compList[{index}].structure.atomCenter();\n",
+    "var around = protein.structure.getAtomSetWithinPoint(ligand_center, {radius});\n",
+    "var around_complete = protein.structure.getAtomSetWithinGroup(around);\n",
+    "var last_repr = protein.reprList[protein.reprList.length-1];\n",
+    "protein.removeRepresentation(last_repr);\n",
+    "protein.addRepresentation(\"licorice\", {{sele: around_complete.toSeleString()}});\n",
+    "\"\"\"\n",
+    "\n",
+    "def show_docking(protein, ligands, vina_output):\n",
+    "    # Split the multi PDBQT ligand file into separate files\n",
+    "    ligands_files = split_pdbqt(ligands)\n",
+    "    # Retrieve affinities\n",
+    "    affinities = parse_output(vina_output)['Affinity (kcal/mol)']\n",
+    "    \n",
+    "    # This is the event handler - action taken when the user clicks on the select box\n",
+    "    def _on_selection_change(change):\n",
```

Comment:
The `show_docking` function works and I think it is just great to have that interactive visualization! 
I just don't fully understand why it works :)

The function `_on_selection_change` appears twice - and here even before variable `v` is defined. 
Is this on purpose? If so, I am interested in why :)

Also, could you comment (especially in the talktorial) on why is it necessary and working to define functions within functions? I would have expected to have that `_on_selection_change` function as stand-alone function with `v` as parameter, though I assume that this has something to do with how `ipywidgets.Select` is working.

Reply:
> The function _on_selection_change appears twice - and here even before variable v is defined.

This was a bug, thanks for noticing!

> Also, could you comment (especially in the talktorial) on why is it necessary and working to define functions within functions?

It's about _scope_ and _closures_ (more [info here](https://medium.com/@dannymcwaves/a-python-tutorial-to-understanding-scopes-and-closures-c6a3d3ba0937)). Basically we want `v` (now renamed to `viewer`) to be available in the event callable without having to pass it explicitly (because we can't due to `.observe()` design). Functions will try to locate missing local variables in outer scopes (function scope-> notebook "global" scope-> builtins). Since we don't want it to access the global scope directly (unpredictable behavior in best case), we nest the callable within our main function so it access the function scope first (callable scope -> function scope -> global scope -> builtins).